File: backend/src/models/Assignment.ts
import mongoose, { Schema, type Document } from 'mongoose';

export type AssignmentStatus = 'PENDING' | 'ACCEPTED' | 'DECLINED' | 'EN_ROUTE' | 'IN_PROGRESS' | 'COMPLETED' | 'CANCELLED';

export interface IAssignment extends Document {
  incidentId: mongoose.Types.ObjectId;
  volunteerId: mongoose.Types.ObjectId;
  assignedBy: mongoose.Types.ObjectId;
  status: AssignmentStatus;
  message?: string;
  acceptedAt?: Date;
  declinedAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

const assignmentSchema = new Schema<IAssignment>(
  {
    incidentId: { type: Schema.Types.ObjectId, ref: 'Incident', required: true },
    volunteerId: { type: Schema.Types.ObjectId, ref: 'Volunteer', required: true },
    assignedBy: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    status: {
      type: String,
      enum: ['PENDING', 'ACCEPTED', 'DECLINED', 'EN_ROUTE', 'IN_PROGRESS', 'COMPLETED', 'CANCELLED'],
      default: 'PENDING'
    },
    message: { type: String, trim: true },
    acceptedAt: { type: Date },
    declinedAt: { type: Date }
  },
  { timestamps: true }
);

assignmentSchema.index({ incidentId: 1, status: 1 });
assignmentSchema.index({ volunteerId: 1, status: 1 });
assignmentSchema.index({ assignedBy: 1, createdAt: -1 });

export const Assignment = mongoose.model<IAssignment>('Assignment', assignmentSchema);

File: backend/src/main/java/com/raava/model/Assignment.java
package com.raava.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;

@Document(collection = "assignments")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Assignment {
    @Id
    private String id;

    @DBRef(lazy = true)
    private Incident incident;

    @DBRef(lazy = true)
    private Volunteer volunteer;

    @DBRef(lazy = true)
    private User assignedBy;

    private Status status = Status.PENDING;
    private String message;

    private Instant createdAt;
    private Instant updatedAt;

    public enum Status { PENDING, ACCEPTED, DECLINED, EN_ROUTE, IN_PROGRESS, COMPLETED, CANCELLED }
}
